posts/2021-03-13-microprofile-config-2.0.adoc

```diff
@@ -240,7 +254,7 @@ The behaviour for "empty" and "special" config property values has been updated:
 * A value is considered to be "empty" if the link:https://download.eclipse.org/microprofile/microprofile-config-2.0/apidocs/org/eclipse/microprofile/config/spi/Converter.html[Converter] being used considers it to be "empty". For example: 
 ** All Converters consider `""`, the empty String, to be empty.
 ** The built-in Converter for String[] considers `","` to be empty (because it is "special"). 
-* Retrieving these "empty" values "natively" will now throw a `NoSuchElementException`, e.g. for the values defined in the <<example-ConfigSource2, example ConfigSource v2>>:
+* From MicroProfile Config 2.0, these "empty" values are no longer valid. Retrieving the values "natively" (without defaultValues or Optionals) will now throw a `NoSuchElementException`. E.g. for the values defined in the <<example-ConfigSource2, example ConfigSource v2>>:
```

Comment:
```suggestion
* From MicroProfile Config 2.0 onwards, these "empty" values are no longer valid. Retrieving the values without using an Optional or a default value will now throw a `NoSuchElementException`. E.g. for the values defined in the <<example-ConfigSource2, example ConfigSource v2>>:
```

Reply:
I think we should probably abort "or a default value" since defining a `defaultValue` is only used if the value "[does not exist](https://download.eclipse.org/microprofile/microprofile-config-2.0/apidocs/org/eclipse/microprofile/config/inject/ConfigProperty.html#defaultValue--)", which is _(I think)_ different to "empty".

For example, with `empty.string.array=,` defined in a ConfigSource, 
```
    @Inject
    @ConfigProperty(name = "empty.string.array", defaultValue = "a default value")
    Optional<String[]> emptyStringArray;
```
`emptyStringArray` still returns `Optional.empty`

Similarly, 
```
    @Inject
    @ConfigProperty(name = "empty.string.array", defaultValue = "a default value")
    String[] emptyStringArray;
```
Still throws a `DeploymentException` (caused by a `NoSuchElementException`)

Reply:
"empty" is now the same as "does not exist"

Reply:
are you sure about that second example? providing a defaultValue should mean it does NOT throw an exception

Reply:
From the conclusion of [this issue](https://github.com/eclipse/microprofile-config/issues/608):

> throw NoSuchElementException when converter returns null. DefaultValue is only a lowest config source.

where returning null is being interpreted as "empty"

Reply:
The only related TCK I can find is [EmptyValuesTest.java](https://github.com/eclipse/microprofile-config/blob/master/tck/src/main/java/org/eclipse/microprofile/config/tck/emptyvalue/EmptyValuesTest.java) which I think is just testing that the Injection causes a `DeploymentException` which I think would occur for both of our understandings since the `defaultValue` is defined as `""`.

Reply:
yes, defaultValue="" would be the same as not having a default value
but, defaultValue="anything" should never result in a NoSuchElementException (for type String[] ... some weird custom Converter<MyType> could still return a null)

Reply:
What makes you say adding "defaultValue="anything" should never result in a NoSuchElementException"? I agree this sounds logical, but I don't think this is what the spec says, and [this discussion](https://github.com/eclipse/microprofile-config/issues/608)

Reply:
String[] is a built in converter which first splits on ",". String is an automatic converter. For the input "anything", the output would be ["anything"] (non-null). Why do you think it could be null (and thus result in a NSEE)?